### ISSUE TITLE
FEATURE REQUEST: Resource Requests/Requirements in moby build yml  

### ISSUE THREAD

**Issue by ndauten** [linuxkit/linuxkit#2155]
**Description**

The tool I have, memorizer (shortly to be added via PR to projects), requires a certain amount of RAM and also performs better with fewer cores. What is the best way to specify these constraints? For example, can we add a method for attaching metadata to the yml file for the moby build? 

@deitch and I had a minor discussion in slack on the topic and decided to open a more formal discussion here.

**Comment by deitch** [comment]
We did, indeed (thanks for opening the issue @ndauten ).

To the best of my knowledge, there is no way for the final output formats to specify constraints like "minimum RAM" etc. There certainly isn't any standard for it. 

However, this may vary by output format (gcp, aws, etc.). If there are ways to specify it, I could see how we could add to to the `.yml` and then attach it to the output image (if it supports it), but I don't know if any such exist.

**Comment by justincormack** [comment]
We should probably have a generic metadata section. It is not clear how you
could attach this to built images in a way that can be usefully extracted,
eg an AMI has no metadata support that I am aware of.

**Comment by deitch** [comment]
Yeah. I think in theory this is a good idea. But I wouldn't want to put metadata about resource requirements in the yml that never makes it to the final image; I feel like that would be misleading and lead to lots of "i required 2GB of mem, why iddint it work?"

**Comment by ndauten** [comment]
So let me make sure I understand. The image is something like a statically linked library? 

And the backend "run" driver has hard coded data in LinuxKit src. I see now what @deitch meant about the image being a how to build vs how to run issue. 

I agree that people might get annoyed by the constraints deal, and it isn't terrible to just punt on this and assume the user should setup their own proper command to linuxkit, but it would be nice to convey these constraints to external entities. Two primary ones: other uses (could be done in a README?) and the automated build systems. So if I do a PR to LinuxKit and my system requires a lot of memory, then someone has to configure the CI specifically to handle that project? Not a bad way to do it.

**Comment by deitch** [comment]
> The image is something like a statically linked library?

Not exactly. The "image" is a bootable linux image. It can be given in any of several formats that are accepted by various boot systems: AWS AMI, kernel+initrd (gzipped cpio), EFI-bootable iso, etc.

In the end, all of those are just wrapping the 2-3 key files: kernel, initrd, cmdline options. It it just doing it in different ways to match the launch system (AWS vs iso runner vs GCP vs VMWare vs etc.).

> but it would be nice to convey these constraints to external entities

I agree wholeheartedly. And adding the metadata to the yml - as long as it can be added to the image in a useful way - is a really good idea. But I don't know of any standard that is mutually agreed for conveying, "I am an image and recommend minimum resources of x/y/z." (pronounced "zed" in my book :-) )

That having been said, if there is none, and we could set some reasonable standard, we could begin such a process. Where would one store it though?

**Comment by rn** [comment]
well, there is [OVF](https://en.wikipedia.org/wiki/Open_Virtualization_Format) which probably allows you to specify this sort of stuff but it contains way too much XML for my liking
